drivers/mcp2515/mcp2515_spi.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) 2016 OTA keys
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mcp2515
+ * @{
+ *
+ * @file
+ * @brief       MCP2515 can spi driver
+ *
+ * @author      Toon Stegen <toon.stegen@altran.com>
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "mcp2515_spi.h"
+#include "mcp2515_defines.h"
+
+#include "periph/gpio.h"
+#include "periph/spi.h"
+
+#include "xtimer.h"
+#include "irq.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+int mcp2515_spi_init(const candev_mcp2515_t *dev)
+{
+    int res;
+
+    /* Configure SPI */
+    res = spi_init_cs(dev->conf->spi, dev->conf->cs_pin);
+    if (res != SPI_OK) {
+        DEBUG("spi_init_master: error initializing SPI_%i device (code %i)\n",
+              dev->conf->spi, res);
+        return -1;
+    }
+    return 0;
+}
+
+int mcp2515_spi_reset(const candev_mcp2515_t *dev)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, false,
+                      MCP2515_SPI_RESET);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read(const candev_mcp2515_t *dev, uint8_t addr, uint8_t *buf,
+                     unsigned int len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ);
+    spi_transfer_regs(dev->conf->spi, dev->conf->cs_pin, addr, NULL,
+                      (void *)buf, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_read_rxbuf(const candev_mcp2515_t *dev, uint8_t mailbox,
+                           void *buf, uint8_t len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
+                dev->conf->spi_clk);
+    spi_transfer_byte(dev->conf->spi, dev->conf->cs_pin, true,
+                      MCP2515_SPI_READ_RXBUF | (mailbox << 2));
+    spi_transfer_bytes(dev->conf->spi, dev->conf->cs_pin, false, NULL,
+                       (void *)buf, len);
+    spi_release(dev->conf->spi);
+    return 0;
+}
+
+int mcp2515_spi_write(const candev_mcp2515_t *dev, uint8_t addr, uint8_t *buf,
+                      unsigned int len)
+{
+    spi_acquire(dev->conf->spi, dev->conf->cs_pin, dev->conf->spi_mode,
```

Comment:
I'm currently running extensive tests where two boards continuously send to each other. I noticed that after several hours, one of the boards always seemed to freeze. My guess is that this has to do with this `spi_acquire`. I think that when an spi-write has started, and at that very moment an rx-interrupt is received, the controller enters a deadlock. 

I was able to fix this by disabling interrupts before writing and re-enabling them afterwards.

I have a patch ready for this, but am still investigating a very similar problem that now has arisen on one of the boards. I want to be sure that the problem is solved before I push patches :)

Reply:
Mhmmm, I think I fixed this in 64db71635e19bbd25bd4bc27078c6868b6e26685. Back then I was working on a project that used this driver. I noticed that the CAN-system ended up in a deadlock, and thus froze the whole program. 
I remember spending quite some time on finding the cause, but since everything works in that project right now, I think it is fixed. (If I remember correctly, it had something to do with an interrupt preventing an spi_release from happening. The next time we would send something, the spi_aquire would block indefenitely).